src/streaming/zarr.common.cpp

```diff
@@ -0,0 +1,297 @@
+#include "macros.hh"
+#include "zarr.common.hh"
+
+#include <stdexcept>
+
+size_t
+zarr::bytes_of_type(ZarrDataType data_type)
+{
+    switch (data_type) {
+        case ZarrDataType_int8:
+        case ZarrDataType_uint8:
+            return 1;
+        case ZarrDataType_int16:
+        case ZarrDataType_uint16:
+            return 2;
+        case ZarrDataType_int32:
+        case ZarrDataType_uint32:
+        case ZarrDataType_float32:
+            return 4;
+        case ZarrDataType_int64:
+        case ZarrDataType_uint64:
+        case ZarrDataType_float64:
+            return 8;
+        default:
+            throw std::runtime_error("Invalid data type: " +
+                                     std::to_string(data_type));
+    }
+}
+
+size_t
+zarr::bytes_of_frame(const std::vector<Dimension>& dims, ZarrDataType type)
+{
+    return bytes_of_type(type) * dims.back().array_size_px *
+           dims[dims.size() - 2].array_size_px;
```

Comment:
What do the last and second-to-last elements represent? Assigning these values to named variables could enhance clarity. For instance, if they represent width and height, you might rewrite it as:
```cpp
const auto width = dims[dims.size() - 1].array_size_px;
const auto height = dims[dims.size() - 2].array_size_px;
return bytes_of_type(type) * width * height;
```